### ISSUE TITLE
JSX import fixes do not support new React JSX Transform (without react import)

### ISSUE THREAD

**Issue by cduff** [microsoft/TypeScript#41762]
- VSCode Version: 1.51.1
- OS Version: Windows 10 Pro

React 17 introduced a [new JSX transform](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html) which does not require react to be imported in every file. Support for this was added in [TypeScript 4.1](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#jsx-factories). However vscode is still suggesting/requiring the react import before other JSX import fixes are suggested.

Steps to Reproduce:

1. `npx create-react-app my-app --template typescript`
2. `cd my-app`
3. `code --disable-extensions .`
4. Tell vscode to [use the Workspace TypeScript version](https://code.visualstudio.com/docs/typescript/typescript-compiling#_using-newer-typescript-versions) (4.1.2).
5. Add a `src/Test.tsx` file with content `export default function Test() { return <App />; }`. It should show a problem `Cannot find name 'App'.`.
6. Click to show fixes. It unnecessarily suggests/requires `import default 'React' from module "react"` before it will suggest the desired `import default 'App' from module "./App"`.

In short, you can't get help with JSX imports without littering your files with unnecessary react imports.

Does this issue occur when all extensions are disabled?: Yes

**Comment by RyanCavanaugh** [comment]
We're still working on exactly figuring out how the new helpful automatic import behavior should interact with TS's framework-agnostic view of JSX

**Comment by kuasha420** [comment]
Any progress made on this yet?

**Comment by nahtnam** [comment]
One workaround I found is: Instead of pressing `cmd+.` to open the auto suggest panel, remove the last letter of the component and type it back. You should see the suggest panel come up with the component name instead of the `import React from 'react'` suggestion.

**Comment by PythonCreator27** [comment]
I tried this workaround too, but when I am trying to fix all missing imports, this is a very annoying task. Instead, I `import React from 'react';` and then fix all missing imports, and then it is as simple as removing the react import.

**Comment by kuasha420** [comment]
> I tried this workaround too, but when I am trying to fix all missing imports, this is a very annoying task. Instead, I `import React from 'react';` and then fix all missing imports, and then it is as simple as removing the react import.

I have a precommit hook that runs some modified version of that react-jsx codemod as part of lint staged. VERY UGLY solution but that's the best I could do atm. :/

**Comment by Paden** [comment]
> We're still working on exactly figuring out how the new helpful automatic import behavior should interact with TS's framework-agnostic view of JSX

Since JSX is no longer apart of a single framework, does this still apply?

**Comment by cmlenz** [comment]
I think this is the same as #41920

**Comment by jmadelaine** [comment]
Any updates on this?

**Comment by zackdotcomputer** [comment]
@cmlenz If it is the same then I might have bad news because I pulled in the nightly with #41950 (the fix for #41920) and this issue is still present :-\

**Comment by Lycolia** [comment]
I try  TypeScript 4.3.2, but still not solved it. why is that?
This used VSCode Version 1.56.2.

**Comment by lucasousi** [comment]
Adding @zackdotcomputer answer, remember to configure the typescript version in your VSCODE

**Comment by zenVentzi** [comment]
![image](https://user-images.githubusercontent.com/11378653/121815056-00b87900-cc7d-11eb-9a9f-90bec89f38a1.png)

Am I missing anything? Using 4.3.2 and still happening

**Comment by lexanth** [comment]
As noted in [this comment](https://github.com/microsoft/TypeScript/issues/41466#issuecomment-818975689) it only works if you have `"jsx": "react-jsx"` or `"jsx": "react-jsxdev"` in your tsconfig - it doesn't work if you're using `"jsx": "preserve"`.

I'm not sure how you'd get around that - ts would need some other way to work out that you're using the new transform.

**Comment by eschwartz** [comment]
So glad to have found this! And thank you @lexanth for the fix.

Just to clarify for anyone else stumbling on this thread, you need to a a `jsconfig.json` file to your project root:

```json
{
  "compilerOptions": {
    "baseUrl": "./src",
    "checkJs": true,
    "jsx": "react-jsx"
  }
}
```

You can then use `CMD + .` to fix a missing import:

![2021-06-18 10 30 39](https://user-images.githubusercontent.com/1153371/122585077-b2123280-d020-11eb-822c-731d01908981.gif)
![2021-06-18 10 31 06](https://user-images.githubusercontent.com/1153371/122585080-b3435f80-d020-11eb-8617-84b688287cd1.gif)

This has been driving me nuts forever, so glad to have found a solution here!

**Comment by Elia-Darwish** [comment]
Maybe it’s also worth mentioning for those who are using Naxt.js that this solution will not work for them  . Next.js uses `"jsx": "preserve"`

**Comment by PythonCreator27** [comment]
Is there any solution or workaround for next.js users?

**Comment by nghiepit** [comment]
Any update for Next.js? Using 4.3.4 and still happening 😭

**Comment by eliasjnior** [comment]
> Any update for Next.js? Using 4.3.4 and still happening 😭

For now, what I did is remove `"react/react-in-jsx-scope": "off",` from my eslint :/

**Comment by tylerjlawson** [comment]
@nghiepit does v4.3.5 resolve the issue? 

#44724

**Comment by nghiepit** [comment]
@tylerjlawson I'm working on v4.3.5 and the problem is still.
Am I missing something?

![image](https://user-images.githubusercontent.com/4768095/126358197-06324cf0-7378-4b0e-a2a6-ff9a6e31d156.png)

**Comment by PythonCreator27** [comment]
@nghiepit Are you using next.js?

**Comment by nghiepit** [comment]
@PythonCreator27 Yes. I'm using Next.js
This is my `tsconfig.json` and `VSCode 1.58.2`
```json
{
  "compilerOptions": {
    "target": "es5",
    "lib": ["dom", "dom.iterable", "esnext"],
    "module": "esnext",
    "jsx": "preserve",
    "baseUrl": ".",
    "strict": true,
    "noEmit": true,
    "allowJs": true,
    "skipLibCheck": true,
    "moduleResolution": "node",
    "esModuleInterop": true,
    "isolatedModules": true,
    "resolveJsonModule": true,
    "forceConsistentCasingInFileNames": true
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
  "exclude": ["node_modules", ".next"]
}
```

**Comment by kuasha420** [comment]
`"jsx": "preserve",` will not work here. But nextjs won't work without preserve. I think Nextjs needs to fix this on their side.

**Comment by Admiralfeb** [comment]
Any ideas why the TS engine doesn't pick up react if you set jsx to "preserve" or the reasoning behind the decision?

The handbook on it says it only affects the emit stage and does not affect type-checking.

If I'm working in other configurations and React is at v17, then I shouldn't have to have React imported in each file. If I have my own transpilation system, such as Next, Babel, etc. I'd still like to not have VSCode complaining that React isn't imported into the file.

**Comment by Shahaed** [comment]
Would love a response from the Typescript team on this

**Comment by andrewbranch** [comment]
Looking into it 👍

**Comment by pvinis** [comment]
what about `"jsx":"react-native"`? should there be a `"jsx":"react-native-jsx"` too?
